File: src/test/java/cube8540/oauth/authentication/credentials/oauth/client/infra/rule/DefaultClientIdValidationRuleTest.java
package cube8540.oauth.authentication.credentials.oauth.client.infra.rule;

import cube8540.oauth.authentication.credentials.oauth.client.domain.OAuth2Client;
import cube8540.oauth.authentication.credentials.oauth.client.domain.OAuth2ClientId;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@DisplayName("기본 클라이언트 아이디 유효성 검사 클래스 테스트")
class DefaultClientIdValidationRuleTest {

    private static final String RAW_CLIENT_LESS_THEN_ALLOWED_LENGTH = "oauth";
    private static final OAuth2ClientId CLIENT_LESS_THEN_ALLOWED_LENGTH = new OAuth2ClientId(RAW_CLIENT_LESS_THEN_ALLOWED_LENGTH);
    private static final String RAW_CLIENT_GRATER_THEN_ALLOWED_LENGTH = "oauth-client-id-oauth-client-id-id";
    private static final OAuth2ClientId CLIENT_GRATER_THEN_ALLOWED_LENGTH = new OAuth2ClientId(RAW_CLIENT_GRATER_THEN_ALLOWED_LENGTH);
    private static final String RAW_CLIENT_WITH_NOT_ALLOWED_SPECIAL_CHARACTER = "oauth-client-id#";
    private static final OAuth2ClientId CLIENT_WITH_NOT_ALLOWED_SPECIAL_CHARACTER = new OAuth2ClientId(RAW_CLIENT_WITH_NOT_ALLOWED_SPECIAL_CHARACTER);

    private DefaultClientIdValidationRule rule;

    @BeforeEach
    void setup() {
        this.rule = new DefaultClientIdValidationRule();
    }

    @Test
    @DisplayName("클라이언트 아이디가 허용되는 길이보다 짧을시 유효성 검사 결과는 false 가 반환되어야 한다.")
    void clientIdLessThenAllowedLengthValidationShouldReturnsFalse() {
        OAuth2Client client = mock(OAuth2Client.class);

        when(client.getClientId()).thenReturn(CLIENT_LESS_THEN_ALLOWED_LENGTH);

        assertFalse(rule.isValid(client));
    }

    @Test
    @DisplayName("클라이언트 이이디가 허용되는 길이보다 길시 유효성 검사 결과는 false 가 반환된어야 한다.")
    void clientIdGraterThenAllowedLengthValidationShouldReturnsFalse() {
        OAuth2Client client = mock(OAuth2Client.class);

        when(client.getClientId()).thenReturn(CLIENT_GRATER_THEN_ALLOWED_LENGTH);

        assertFalse(rule.isValid(client));
    }

    @Test
    @DisplayName("클라이언트 아이디에 허용되지 않는 특수문자가 포함되어 있을시 유효성 검사 결과는 false 가 반환되어야 한다.")
    void clientIdIncludeNotAllowedSpacialCharacterValidationShouldReturnsFalse() {
        OAuth2Client client = mock(OAuth2Client.class);

        when(client.getClientId()).thenReturn(CLIENT_WITH_NOT_ALLOWED_SPECIAL_CHARACTER);

        assertFalse(rule.isValid(client));
    }
}

File: src/test/kotlin/cube8540/oauth/authentication/oauth/client/infra/DefaultClientIdValidationRuleTest.kt
package cube8540.oauth.authentication.oauth.client.infra

import cube8540.oauth.authentication.oauth.client.domain.OAuth2Client
import cube8540.oauth.authentication.oauth.client.domain.OAuth2ClientId
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import java.util.stream.Stream

class DefaultClientIdValidationRuleTest {
    private val minClientIdLength = 8
    private val maxClientIdLength = 30

    private val charPool: List<Char> = ('a'..'z') + ('0'..'9') + listOf('-', '_')
    private val client: OAuth2Client = mockk()

    private val rule = DefaultClientIdValidationRule()

    @ParameterizedTest
    @MethodSource(value = ["minLengthClientIdProvider", "maxLengthClientIdProvider"])
    fun `client id invalid length`(clientId: OAuth2ClientId) {
        every { client.clientId } returns clientId

        val result = rule.isValid(client)
        assertThat(result).isFalse
    }

    @Test
    fun `client id include not allowed special character`() {
        val clientId = OAuth2ClientId("clientId!@#")

        every { client.clientId } returns clientId

        val result = rule.isValid(client)
        assertThat(result).isFalse
    }

    private fun minLengthClientIdProvider(): Stream<Arguments> {
        val results = ArrayList<OAuth2ClientId>()
        for (i in 1 until minClientIdLength) {
            val clientId = (1..i)
                .map { charPool.random() }
                .joinToString("")
            results.add(OAuth2ClientId(clientId))
        }
        return results.map { Arguments.of(it) }.stream()
    }

    private fun maxLengthClientIdProvider(): Stream<Arguments> {
        val clientId = (1..(maxClientIdLength + 1))
            .map { charPool.random() }
            .joinToString("")
        return Stream.of(Arguments.of(OAuth2ClientId(clientId)))
    }
}